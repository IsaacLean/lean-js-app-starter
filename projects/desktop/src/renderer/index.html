<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <%= csp %>
    <title>Lean JavaScript Application Starter</title>
  </head>
  <body>
    <div id="root"></div>
    <noscript>You must enable JavaScript to use this application!</noscript>
  </body>
</html>