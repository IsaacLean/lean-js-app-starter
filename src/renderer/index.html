<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>Lean JavaScript Application Starter</title>
  </head>
  <body>
    <div id="root"></div>
    <noscript>You must enable JavaScript to use this application!</noscript>
  </body>
</html>