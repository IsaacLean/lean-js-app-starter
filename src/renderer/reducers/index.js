// @flow
import { combineReducers } from 'redux'

import byId, { getTodo } from './byId'
import createList, { getIds } from './createList'
import type { State } from '../types'

const listByFilter = combineReducers({
  all: createList('all'),
  active: createList('active'),
  completed: createList('completed')
})

const todos = combineReducers({
  byId,
  listByFilter
})

export default todos

export const getVisibleTodos = (state: State, filter: string) => {
  if(state.listByFilter && state.byId) {
    const ids = getIds(state.listByFilter[filter])
    if(ids) {
      return ids.map(id => {
        if(state.byId) return getTodo(state.byId, id)
      })
    }
  }
}
