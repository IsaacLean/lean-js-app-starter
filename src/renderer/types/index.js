// @flow
import type { Dispatch as ReduxDispatch } from 'redux'

export type Action =
  | Action_AddTodoSuccess
  | Action_FetchTodosRequest
  | Action_FetchTodosSuccess
  | Action_ToggleTodoSuccess

export type Action_AddTodoSuccess = {|
  type: 'ADD_TODO_SUCCESS',
  response: NormalizedRes
|}

export type Action_FetchTodosRequest = {|
  type: 'FETCH_TODOS_REQUEST',
  filter: string
|}

export type Action_FetchTodosSuccess = {|
  type: 'FETCH_TODOS_SUCCESS',
  filter: string,
  response: NormalizedRes
|}

export type Action_ToggleTodoSuccess = {|
  type: 'TOGGLE_TODO_SUCCESS',
  response: NormalizedRes
|}

export type Dispatch = ReduxDispatch<Action> & (action: ThunkAction) => any

export type GetState = () => State

export type NormalizedRes = {
  entities: {
    [string]: {
      [string]: any
    }
  },
  result: string | Array<string>
}

export type State = {|
  +byId: State_ById,
  +listByFilter: State_ListByFilter
|} | {||}

export type State_ById = {
  +[string]: Todo
} | {||}

export type State_ListByFilter = {|
  +all: State_List
|} | {||}

export type State_List = {
  +ids: State_List_Ids
} | {||}

export type State_List_Ids = Array<string>

export type ThunkAction = (dispatch: Dispatch, getState: GetState) => any

export type Todo = {
  id: string,
  text: string,
  completed: boolean
}
