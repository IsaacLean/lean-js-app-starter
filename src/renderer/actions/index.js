// @flow
import { normalize } from 'normalizr'

import * as api from '../util/mockAPI'
import { arrayOfTodos, todo } from '../types/schema'
import type { Dispatch, ThunkAction, Todo } from '../types'

export const addTodo = (text: string) => (dispatch: Dispatch) =>
  api.addTodo(text).then(response => dispatch({
    type: 'ADD_TODO_SUCCESS',
    response: normalize(response, todo)
  }))

export const fetchTodos = (filter: string): ThunkAction => (dispatch: Dispatch) => {
  dispatch({
    type: 'FETCH_TODOS_REQUEST',
    filter
  })

  return api.fetchTodos(filter)
    .then(response => dispatch(fetchTodosSuccess(filter, response)))
}

export const fetchTodosSuccess = (filter: string, response: Array<Todo>) => ({
  type: 'FETCH_TODOS_SUCCESS',
  filter,
  response: normalize(response, arrayOfTodos)
})

export const toggleTodo = (id: string) => (dispatch: Dispatch) =>
  api.toggleTodo(id).then(response => dispatch({
    type: 'TOGGLE_TODO_SUCCESS',
    response: normalize(response, todo)
  }))
