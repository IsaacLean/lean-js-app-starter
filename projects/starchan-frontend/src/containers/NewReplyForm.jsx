import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'

import { createReply, setErr } from '../actions'

class NewReplyFormClass extends Component {
  state = {
    comment: '',
    show: false
  }

  handleChange = (e, state) => {
    this.setState({ [state]: e.currentTarget.value })
  }

  handleClick = e => {
    e.preventDefault()
    this.setState({ show: true })
  }

  handleSubmit = e => {
    e.preventDefault()
    if(this.props.createReply) {
      this.props.createReply(
        this.props.match.params.id,
        { comment: this.state.comment }
      )
        .then(() => {
          this.reset()
        })
        .catch(err => {
          if(this.props.setErr) {
            this.props.setErr('create', err)
          }
        })
    }
  }

  reset = () => {
    this.setState({
      comment: '',
      show: false
    })
  }

  render() {
    if(this.state.show) {
      let err
      if(this.props.err) {
        let statusTxt
        if(this.props.err.res && this.props.err.res.status) statusTxt = `${this.props.err.res.status} - `

        err = (
          <tr>
            <td colSpan="2">
              <b className="center">ERROR: {statusTxt}{this.props.err.message}</b>
            </td>
          </tr>
        )
      }

      return <form id="new-form" className="center" onSubmit={this.handleSubmit}>
        <table>
          <tbody>
            <tr>
              <th><label htmlFor="comment">Comment</label></th>
              <td>
                <textarea
                  id="comment"
                  value={this.state.comment}
                  onChange={e => this.handleChange(e, 'comment')}
                  required
                />
              </td>
            </tr>
            {err}
            <tr>
              <td colSpan="2">
                <button type="submit">Post</button>
              </td>
            </tr>
          </tbody>
        </table>
      </form>
    }
    return <span className="center">[<a href="#" onClick={this.handleClick}>Post a Reply</a>]</span>
  }
}

NewReplyFormClass.propTypes = {
  createReply: PropTypes.func,
  err: PropTypes.object,
  match: PropTypes.object.isRequired,
  setErr: PropTypes.func
}

export const NewReplyForm = withRouter(NewReplyFormClass)

const mapStateToProps = state => ({
  err: state.err.create
})

const mapDispatchToProps = dispatch => bindActionCreators({ createReply, setErr }, dispatch)

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(NewReplyFormClass))
