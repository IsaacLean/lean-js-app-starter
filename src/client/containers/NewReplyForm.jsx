import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'

import { createReply, setErr } from '../actions'

class NewReplyForm extends Component {
  state = {
    comment: '',
    show: false
  }

  handleChange = (e, state) => {
    this.setState({ [state]: e.currentTarget.value })
  }

  handleClick = e => {
    e.preventDefault()
    this.setState({ show: true })
  }

  handleSubmit = e => {
    e.preventDefault()
    this.props.createReply(
      this.props.match.params.id,
      { comment: this.state.comment }
    )
      .then(() => {
        this.reset()
      })
      .catch(err => {
        this.props.setErr('create', err)
      })
  }

  reset = () => {
    this.setState({
      comment: '',
      show: false
    })
  }

  render() {
    if(this.state.show) {
      let err
      if(this.props.err) {
        let statusTxt
        if(this.props.err.res && this.props.err.res.status) statusTxt = `${this.props.err.res.status} - `

        err = (
          <tr>
            <td colSpan="2">
              <b className="center">ERROR: {statusTxt}{this.props.err.message}</b>
            </td>
          </tr>
        )
      }

      return <form id="new-thread-form" className="center" onSubmit={this.handleSubmit}>
        <table>
          <tbody>
            <tr>
              <th><label htmlFor="comment">Comment</label></th>
              <td>
                <textarea
                  id="comment"
                  value={this.state.comment}
                  onChange={e => this.handleChange(e, 'comment')}
                  required
                />
              </td>
            </tr>
            {err}
            <tr>
              <td colSpan="2">
                <button type="submit">Post</button>
              </td>
            </tr>
          </tbody>
        </table>
      </form>
    }
    return <span className="center">[<a href="#" onClick={this.handleClick}>Post a Reply</a>]</span>
  }
}

NewReplyForm.propTypes = {
  createReply: PropTypes.func.isRequired,
  err: PropTypes.object,
  match: PropTypes.object.isRequired,
  setErr: PropTypes.func.isRequired
}

const mapStateToProps = state => ({
  err: state.err.create
})

const mapDispatchToProps = dispatch => bindActionCreators({ createReply, setErr }, dispatch)

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(NewReplyForm))
