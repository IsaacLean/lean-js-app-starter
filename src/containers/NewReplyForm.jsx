import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'

import { createReply } from '../actions'

class NewReplyForm extends Component {
  state = {
    comment: '',
    show: false
  }

  handleChange = (e, state) => {
    this.setState({ [state]: e.currentTarget.value })
  }

  handleClick = e => {
    e.preventDefault()
    this.setState({ show: true })
  }

  handleSubmit = e => {
    e.preventDefault()
    this.props.createReply(
      this.props.match.params.id,
      { comment: this.state.comment }
    )
      .then(thread => {
        this.setState({ show: false })
        this.reset()
      })
  }

  reset = () => {
    this.setState({ comment: '' })
  }

  render() {
    if(this.state.show) {
      return <form id="new-thread-form" className="center" onSubmit={this.handleSubmit}>
        <table>
          <tbody>
            <tr>
              <th><label htmlFor="comment">Comment</label></th>
              <td>
                <textarea
                  id="comment"
                  value={this.state.comment}
                  onChange={e => this.handleChange(e, 'comment')}
                  required
                />
              </td>
            </tr>
            <tr>
              <td colSpan="2">
                <button type="submit">Post</button>
              </td>
            </tr>
          </tbody>
        </table>
      </form>
    }
    return <span className="center">[<a href="#" onClick={this.handleClick}>Post a Reply</a>]</span>
  }
}

NewReplyForm.propTypes = {
  createReply: PropTypes.func.isRequired,
  match: PropTypes.object.isRequired
}

const mapDispatchToProps = dispatch => bindActionCreators({ createReply }, dispatch)

export default withRouter(connect(null, mapDispatchToProps)(NewReplyForm))
