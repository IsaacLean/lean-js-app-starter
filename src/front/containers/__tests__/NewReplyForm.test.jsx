import React from 'react'
import renderer from 'react-test-renderer'
import { HashRouter as Router } from 'react-router-dom'

import { NewReplyForm } from '../NewReplyForm'

describe('NewReplyForm', () => {
  it('renders properly', () => {
    const component = renderer.create(
      <Router>
        <NewReplyForm />
      </Router>
    )
    expect(component).toMatchSnapshot()
  })
})
