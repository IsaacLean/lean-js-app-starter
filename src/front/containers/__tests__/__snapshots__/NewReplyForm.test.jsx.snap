// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewReplyForm renders properly 1`] = `
<span
  className="center"
>
  [
  <a
    href="#"
    onClick={[Function]}
  >
    Post a Reply
  </a>
  ]
</span>
`;
